gossip/gossip.go

```diff
@@ -0,0 +1,109 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package gossip
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ava-labs/avalanchego/codec"
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ava-labs/coreth/peer"
+	"github.com/ava-labs/coreth/plugin/evm/message"
+)
+
+type Handler[T Tx] interface {
+	GetMempool() Mempool[T]
+	NewGossipRequest() message.MempoolTxsRequest
+	NewTx() T
+}
+
+func NewPullGossiper[T Tx](
+	handler Handler[T],
+	networkClient peer.NetworkClient,
+	codec codec.Manager,
+	gossipFrequency time.Duration,
+	gossipSize int,
+) *PullGossiper[T] {
+	return &PullGossiper[T]{
+		handler:         handler,
+		client:          networkClient,
+		codec:           codec,
+		gossipFrequency: gossipFrequency,
+		gossipSize:      gossipSize,
+	}
+}
+
+type PullGossiper[T Tx] struct {
+	handler         Handler[T]
+	client          peer.NetworkClient
+	codec           codec.Manager
+	gossipFrequency time.Duration
+	gossipSize      int
+}
+
+func (p *PullGossiper[T]) Start(shutdown chan struct{}, shutdownWg *sync.WaitGroup) {
+	gossipTicker := time.NewTicker(p.gossipFrequency)
+
+	defer func() {
+		gossipTicker.Stop()
+		shutdownWg.Done()
+	}()
+
+	for {
+		select {
+		case <-gossipTicker.C:
+			peers, err := p.client.Sample(p.gossipSize)
+			if err != nil {
+				log.Error("failed to sample peers", "error", err)
+				return
+			}
+
+			for nodeID := range peers {
+				bloom := p.handler.GetMempool().GetPendingTxsBloomFilter()
+				bloomBytes, err := bloom.MarshalBinary()
+				if err != nil {
+					log.Error("failed to marshal bloom filter", "error", err)
+					continue
+				}
+
+				request := p.handler.NewGossipRequest()
+				request.SetBloomFilter(bloomBytes)
+				requestBytes, err := message.RequestToBytes(p.codec, request)
+				if err != nil {
+					log.Error("failed to marshal gossip request", "error", err)
+					continue
+				}
```

Comment:
can we do this once instead of for each nodeID?

Reply:
👍